### ISSUE TITLE
Freeze for a sec and run like a thunderbolt

### ISSUE THREAD

**Issue by PhenoloBoy** [dask/dask#4011]
I'm dealing with a strange problem that I've solved adding a sleeping time and I would like to understand if it's a problem coming from dask or from xarray.
The general contest is that I've to read a dataset, process each single pixel vector (pixel drill) with some functions and put it back in a copy of the original dataset filled with the results.
A pseudo code super condensed would look like this:

cube = xr.open_rasterio(inFilePth)
pixels_pairs = zip(range(len(x), range(len(y))
client = Client()
futures = client.map(pixeldrillcleaner, pixels_pairs, cube=cube)
for future, result in as_completed(futures, with_results=True):
       ts, pixel = result
       cube[dict([(self.x_nm, pixel[0]), (self.y_nm, pixel[1])])] = ts
       cube.to_netcdf(r'c:\temp\test.nc')
 
def pixeldrillcleaner (pixel_pairs, **kwargs):
        cube = kwargs['cube'] 
        time_series = cube[pixel_pairs[0], pixel_pairs[1]].to_series()
        ts = do some pandas time series trick 
        return ts

The point is that if I don't add a sleeping time in the pixeldrillcleaner everything freeze after a first initial start. Seems to me that the process is too fast and the scheduler isn't able to dealing with all the results. 
Is it possible ? or I'm completely wrong 
Have any one had the same problem ?

**Comment by mrocklin** [comment]
This note on dask delayed might be informative.  The same logic applies to
the futures interface:
https://docs.dask.org/en/latest/delayed-best-practices.html#avoid-too-many-tasks

**Comment by PhenoloBoy** [comment]
Thanks for the advice but 2500 tasks seems to me not so many.

**Comment by mrocklin** [comment]
Agreed.  2500 should be fine.  I don't know what's wrong.  I recommend
providing a minimal example that is simple but exhibits the same behavior.

**Comment by PhenoloBoy** [comment]
In a sense you are right but I was already put some pseudocode in it. 
By the way the entire code looks like this:

```
from dask.distributed import Client, as_completed
import xarray as xr
import numpy as np
import nodata
import pandas as pd


def read(path):
    return xr.open_rasterio(path)

def cleaner(pixel_pairs, **kwargs):
        cube = kwargs['cube']

        x = pixel_pairs[0]
        y = pixel_pairs[1]

        ts = cube[dict([('x', x), ('y', y)])].to_series()

        ntnlTs = ts.mask(ts > 250)
        ntnlAgg = ntnlTs.groupby([ntnlTs.index.month, ntnlTs.index.day]).median()

        nanList = ts.where(ts > 250).dropna()

        for index, value in nanList.items():
            
                 /// Do some specific staff ... ///
        
       ts.update(nanList)

        time.sleep(0.5) # <--- without this it stack
    except (RuntimeError,):
        print('error on{}'.format(pixel_pairs))

    return ts, pixel_pairs

def main():
    dataset = read(r'c:\temp\chianti.img')

    x = int(cube.sizes['x'])
    y = int(cube.sizes['y'])

    pixels_pairs = np.indices((x, y)).transpose((1, 2, 0)).reshape((x * y, 2))

    client = Client()
    futures = client.map(nodata.cleaner, pixels_pairs, cube=cube)

    for batch in as_completed(futures, with_results=True).batches():
        for future, result in batch:
            ts, pixel = result
            try:
                cube[dict([('x', pixel[0]), ('y', pixel[1])])] = ts
                cube.to_netcdf(r'c:\temp\test.nc')
            except (RuntimeError,):
                print('error')
```
For all those that whats a full test package at this link ( https://www.dropbox.com/sh/n0gu5j9mqqr5c8l/AADxNKAojPagRCA2v32ITUZqa?dl=0 ) there is a small datasample and all the code needed. With my configuration without the time sleep start properly and get stack in a couple of sec.

**Comment by mrocklin** [comment]
I probably won't spend much time working through this example given the
current size.  You might want to try the xarray mailing list.

http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by PhenoloBoy** [comment]
My apology, reading your blog I realise I've been super "rude". 
In any case as I think that this problem is in the interest of the community I condensed my code and tried to be fulfilled with your suggestions.
The new test doesn't read the data from a file and is affected by the same problem. 
I tested it with Tornado 4.5.3 as well (same result) now I'm back on 5.1.1

`from dask.distributed import Client, as_completed
import xarray as xr
import numpy as np
import pandas as pd
import time

def cleaner(pixel, **kwargs):
    cube = kwargs['cube']
    ts = cube[dict([('x', pixel[0]), ('y', pixel[1])])].to_series().astype(float)
    ts += 5
    time.sleep(0.5)    <-- Without freeze
    return ts, pixel

def main():

    times = pd.date_range('2000-01-01', periods=100)
    x = range(30)
    y = range(30)
    cube = xr.DataArray(np.random.rand(len(times), len(x), len(y)), coords=[times, x, y], dims=['time', 'x', 'y'])

    x = int(cube.sizes['x'])
    y = int(cube.sizes['y'])

    pixels_pairs = np.indices((x, y)).transpose((1, 2, 0)).reshape((x * y, 2))

    client = Client()
    futures = client.map(cleaner, pixels_pairs, cube=cube)

    for future, result in as_completed(futures, with_results=True):
        ts, pixel = result
       cube[dict([('x', pixel[0]), ('y', pixel[1])])] = ts

    print(cube)

if __name__ is '__main__':
    main()
`
I'm working on a windows machine with Python 3.6 x64

**Comment by mrocklin** [comment]
The `cube` data is biggish.  You send it to every task.  Try doing the following before your map call

    cube = client.scatter(cube)

**Comment by PhenoloBoy** [comment]
You are right but even if with the scatter I've the same problem. The only difference is that is blocking at the end of the process instead during it. 
Tomorrow I'll test over a linux machine and I'll let you know.

**Comment by PhenoloBoy** [comment]
Tested over a linux machine and I've got the exact same problem (I had just to stress a little bit more the numbers), this time fortunately I've get back an error output: '[Errno 24] Too many open files'.
The error putted me one more interesting prospective; searching over the issue I found lots similarity with this issue: https://github.com/dask/distributed/issues/1858 
I'll update that issue with mine tests as they used a different approach.
Minor discovery is that under windows machine this error isn't properly managed and didn't appear in the console.

p.s. the scatter didn't help to solve the problem but indeed helped to speed up the client map.
